README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# ukcpR

<!-- badges: start -->

<img src="man/figures/logo.png" align="right" height="139" /> <br>
[![Lifecycle:
experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://lifecycle.r-lib.org/articles/stages.html#experimental)
<!-- badges: end -->

The goal of ukcpR is to make it a bit easier to interact with the UK
[Met Office](https://www.metoffice.gov.uk/)’s [UK Climate
Projections](https://www.metoffice.gov.uk/research/approach/collaboration/ukcp)
(UKCP) data, via its API, then work with the data produced

At the moment the package is **focused on a limited number of document
types, all of which are .csv files**. This may expand to further
document types in the future, but for now this helps in some instances.
Currently you can use this package to run queries on:

- Observations from HadUK-Grid over UK for daily data
- Observations from HadUK-Grid over UK for monthly, seasonal or annual
  data
- Variables from local/regional/global projections over UK for monthly,
  seasonal or annual data

Note: **this package relies on you already having a UKCP account**. You
will not be able to use the API without a key it provides. By default
this package assumes the key will be stored as an environment variable
`UKCP_API_KEY` in your `.Renviron` file, but you can overwrite if
desired. It should be noted that any queries you make via the API will
be logged (and results accessible) in the UKCP web interface.

You also need to have some understanding of possible values for API
parameters. The nature of this package is you can freely enter any old
text, and even if individual parameters are valid, they may not work in
combination. Reading up/doing some trials via the UI initially will
probably save a lot of frustration.

See the [how-to
guide](https://shanej90.github.io/ukcpR/articles/using-ukcpr) for more
information and examples.

## Installation

You can install the development version of ukcpR as follows:

``` r
remotes::install_github("shanej90/ukcpR")
```


NEWS.md
# ukcpR 0.1.0

* Functions to make API calls for certain observational and projected .csv files.
* Download the results of said calls locally.
* Read and do some basic pre-processing of the .csv files in R.
* Coordinate conversion from northings/eastings to latitude/longitude.


tests/testthat.R
# This file is part of the standard setup for testthat.
# It is recommended that you do not modify it.
#
# Where should you do additional test configuration?
# Learn more about the roles of various files in:
# * https://r-pkgs.org/tests.html
# * https://testthat.r-lib.org/reference/test_package.html#special-files

library(testthat)
library(ukcpR)

test_check("ukcpR")


tests/testthat/test_api_call.R
###test that a valid query to the API returns a list
testthat::expect_type(
  start_ukcp_observation_csv_job(
    label = "testtthat-good-query",
    identifier = "LS6_Subset_02",
    area_type = "point",
    area_details = c(292500, 72500),
    collection = "land-obs_5",
    temp_avg = "day",
    year_range = c(1960, 2021),
    variable = "rainfall"
  ),
  "list"
)


tests/testthat/manual_testing.R
#manual testing file due to difficulty of using testthat with API calls that can take minutes to run etc

# devtools::load_all()
#
# #observations csv job--------------------
#
# call_results <-  start_ukcp_observation_csv_job(
#   label = "testtthat-good-query",
#   identifier = "LS6_Subset_02",
#   area_type = "point",
#   area_details = c(292500, 72500),
#   collection = "land-obs_5",
#   temp_avg = "day",
#   year_range = c(1960, 2021),
#   variable = "rainfall"
# )
#
# get_files <- download_ukcp_csv_results(call_results$status_url, output_folder = "assets")
#
#  example_csv <- read_observation_csv("assets/subset_2023-07-03T09-25-13.csv")


R/csv_wrangling.R
#' read_observation_csv
#'
#' Read a UKCP observation data .csv file and handle some basic processing.
#' @param filepath Location of the file.
#' @param convert_coordinates Convert coordinates (if you are using them instead of named area) from northings/eastings to latitude/longitude.
#'
#' @return A dataframe with the processed data (and pertinent metadata) in tall format.
#'
#' @examples
#' \dontrun{
#' read_observation_csv(
#'   "filename.csv"
#')
#'}
#'
#' @export

read_observation_csv <- function(
  filepath,
  convert_coordinates = T
) {

  #error handling---------------------------
  if(!is.logical(convert_coordinates)) {
    stop("`convert_coordinates` must be T/TRUE or F/FLASE (unquoted).")
  }

  if(!is.character(filepath)) {
    stop("`fielpath` must be a string pointing to the desired by-month UKCP .csv file.")
  }

  #get pertinent metadata
  metric <- read.csv(
    filepath,
    stringsAsFactors = F,
    skip = 7,
    nrows = 1,
    header = F
  ) |>
    dplyr::pull(V2)

  units <- stringr::str_extract(metric, "(?<=\\()(.*)(?=\\))")

  time_period <- read.csv(filepath, stringsAsFactors = F, skip = 6, nrow = 1, header = F) |>
    dplyr::pull(V2)

  grid_size <- read.csv(filepath, stringsAsFactors = F, skip = 4, nrow = 1, header = F) |>
    dplyr::pull(V2)

  area <- read.csv(filepath, nrows = 1, skip = 1, stringsAsFactors = F, header = F) |>
    dplyr::pull(V2)

  temporal_avg <- read.csv(filepath, nrows = 1, skip = 5, stringsAsFactors = F, header = F) |>
    dplyr::pull(V2)

  #check area type - is it coordinate based or is it a named area?
  if(stringr::str_detect(grid_size, "grid")) {

    is_grid <- T

  } else {

    is_grid <- F

  }

  #check if data is 'bbox' (ie, there is a grid of readings rather than a table)
  if(stringr::str_count(area, " ") == 3) {

    bbox <- T

  } else {

    bbox <- F

  }

  if(bbox == T) {

    #create main dataframe
    df <- read.csv(
      filepath,
      stringsAsFactors = F,
      skip = 12,
      header = F
      ) |>
      #convert all fields to numeric for ease
      dplyr::mutate(dplyr::across(-V1, as.numeric)) |>
      #format date and fill blanks
      dplyr::mutate(date = strptime(V1, format = "%Y-%m-%d")) |>
      tidyr::fill(date)

    #extract eastings to use as column headings
    col_heads <- unlist(as.character(df[2,]))
    col_heads[1] <- "northings"
    col_heads[length(col_heads)] <- "date"
    colnames(df) <- col_heads

    #put into tall format and filter out redundant rows
    df <- df |>
      dplyr::mutate(
        year = lubridate::year(date),
        month = lubridate::month(date),
        season = dplyr::case_when(
          month %in% c(12, 1, 2) ~ "Winter",
          month %in% c(3:5) ~ "Spring",
          month %in% c(6:8) ~ "Summer",
          month %in% c(9:11) ~ "Autumn"
        )
        ) |>
      dplyr::select(-date) |>
      tidyr::pivot_longer(
        cols = c(-northings) & !dplyr::contains("year") & !dplyr::contains("month"),
        names_to = "eastings",
        values_to = "value"
        ) |>
      dplyr::filter(!is.na(value) & northings != "--") |>
      #finalise columns
      dplyr::select(-temp) |>
      dplyr::transmute(
        year,
        season = ifelse(temporal_avg != "Annual", season, NA_character_),
        month = ifelse(!temporal_avg %in% c("Annual", "Season"), month, NA_character_),
        measurement = metric,
        measurement_units = units,
        period = time_period,
        spatial_rep = grid_size,
        northings = as.numeric(northings),
        eastings = as.numeric(eastings),
        value
        )

    #convert coordinates if requested
    if(convert_coordinates == T) {

      #perform conversion
      converted_coords <- do_coord_conversion(df$northings, df$eastings)

      #add to dataframe
      df <- df |>
        dplyr::mutate(
          northings = converted_coords[["latitude"]],
          eastings = converted_coords[["longitude"]]
        ) |>
        dplyr::rename(
          latitude = northings,
          longitude = eastings
        )

    }

  } else {

    #sort area
    eastings <- stringr::word(area, 1)
    northings <- stringr::word(area, 2)

    #process data
    df <- read.csv(filepath, skip = 10, stringsAsFactors = F) |>
      janitor::clean_names() |>
      dplyr::rename(value = 2) |>
      #format date and fill blanks
      #dplyr::mutate(date = strptime(date, format = "%Y-%m-%d")) |>
      #tidyr::fill(date) |>
      dplyr::mutate(
        year = lubridate::year(date),
        month = lubridate::month(date),
        season = dplyr::case_when(
          month %in% c(12, 1, 2) ~ "Winter",
          month %in% c(3:5) ~ "Spring",
          month %in% c(6:8) ~ "Summer",
          month %in% c(9:11) ~ "Autumn"
        )
      ) |>
      dplyr::transmute(
        year,
        season = ifelse(temporal_avg != "Annual", season, NA_character_),
        month = ifelse(!temporal_avg %in% c("Annual", "Season"), month, NA_character_),
        date = ifelse(temporal_avg == "Daily", date, NA),
        spatial_rep = grid_size,
        measurement = metric,
        measurement_units = units,
        eastings = eastings,
        northings = northings,
        value
      )

    #if convert_coordinates = T, convert from northings/eastings to latitude/longitude
    if(convert_coordinates == T & is_grid == T) {

      converted <- do_coord_conversion(
        df$northings |> as.numeric(),
        df$eastings |> as.numeric()
        )

      df <- df |>
        dplyr::mutate(
          northings = converted[["latitude"]],
          eastings = converted[["longitude"]]
        ) |>
        dplyr::rename(
          latitude = northings,
          longitude = eastings
        )

    }

  }

  return(df)

}


#' read_projection_csv
#'
#' Read a UKCP projection data .csv file(s) and handle some basic processing.
#' @param filepath Location of the file.
#' @param convert_coordinates Convert coordinates from northings/eastings to latitude/longitude.
#'
#' @return A dataframe with the processed data (and pertinent metadata) in tall format.
#'
#' @examples
#' \dontrun{
#' read_projection_csv(
#'   "filename.csv"
#')
#'}
#'
#' @export

read_projection_csv <- function(filepath, convert_coordinates = T) {

  #error handling---------------------------
  if(!is.logical(convert_coordinates)) {
    stop("`convert_coordinates` must be T/TRUE or F/FLASE (unquoted).")
  }

  if(!is.character(filepath)) {
    stop("`fielpath` must be a string pointing to the desired by-month UKCP .csv file.")
  }

  #get pertinent metadata
  metric <- read.csv(
    filepath,
    stringsAsFactors = F,
    skip = 9,
    nrows = 1,
    header = F
  ) |>
    dplyr::pull(V2)

  units <- stringr::str_extract(metric, "(?<=\\()(.*)(?=\\))")

  min_yr <- read.csv(filepath, stringsAsFactors = F, skip = 10, nrow = 1, header = F) |>
    dplyr::pull(V2)

  max_yr <- read.csv(filepath, stringsAsFactors = F, skip = 11, nrow = 1, header = F) |>
    dplyr::pull(V2)

  year_range <- paste0(min_yr, " - ", max_yr)

  grid_size <- read.csv(filepath, stringsAsFactors = F, skip = 6, nrow = 1, header = F) |>
    dplyr::pull(V2)

  model_ids <- read.csv(filepath, stringsAsFactors = F, skip = 2, nrow = 1, header = F) |>
    dplyr::pull(V2)

  warming_scenario <- read.csv(filepath, stringsAsFactors = F, skip = 4, nrow = 1, header = F) |>
    dplyr::pull(V2)

  temporal_avg <- read.csv(filepath, stringsAsFactors = F, skip = 7, nrow = 1, header = F) |>
    dplyr::pull(V2)

    #create main dataframe
    df <- read.csv(
      filepath,
      stringsAsFactors = F,
      skip = 12,
      header = F
    ) |>
      #convert all fields to numeric for ease
      dplyr::mutate(dplyr::across(-V1, as.numeric)) |>
      #format date and fill blanks
      dplyr::mutate(date = strptime(V1, format = "%Y-%m-%d")) |>
      tidyr::fill(date)

    #extract eastings to use as column headings
    col_heads <- unlist(as.character(df[2,]))
    col_heads[1] <- "northings"
    col_heads[length(col_heads)] <- "date"
    colnames(df) <- col_heads

    #put into tall format and filter out redundant rows
    df <- df |>
      dplyr::mutate(
        year = lubridate::year(date),
        month = lubridate::month(date),
        season = dplyr::case_when(
          month %in% c(12, 1, 2) ~ "Winter",
          month %in% c(3:5) ~ "Spring",
          month %in% c(6:8) ~ "Summer",
          month %in% c(9:11) ~ "Autumn"
        )
      ) |>
      dplyr::select(-date) |>
      tidyr::pivot_longer(
        cols = c(-northings) & !dplyr::contains("year") & !dplyr::contains("month"),
        names_to = "eastings",
        values_to = "value"
      ) |>
      dplyr::filter(!is.na(value) & northings != "--") |>
      #finalise columns
      dplyr::select(-temp) |>
      dplyr::mutate(
        year,
        season = ifelse(temporal_avg != "Annual", season, NA_character_),
        month = ifelse(!temporal_avg %in% c("Annual", "Season"), month, NA_character_),
        measurement = metric,
        measurement_units = units,
        period = time_period,
        spatial_rep = grid_size,
        northings = as.numeric(northings),
        eastings = as.numeric(eastings),
        models = model_ids,
        climate_scenario = warming_scenario
      )

    #convert coordinates if requested
    if(convert_coordinates == T) {

      #perform conversion
      converted_coords <- do_coord_conversion(df$northings, df$eastings)

      #add to dataframe
      df <- df |>
        dplyr::mutate(
          northings = converted_coords[["latitude"]],
          eastings = converted_coords[["longitude"]]
        ) |>
        dplyr::rename(
          latitude = northings,
          longitude = eastings
        )

    }


  return(df)

}

#' do_coord_conversion
#'
#' Helper function to convert northings/eastings to latitude/longitude respectively.
#' @param northings Numeric vector of northings to be converted.
#' @param eastings Numeric vector of eastings to be converted.
#'
#' @return A list object containing two vectors, one for latitude and one for longitude.
#'
#' @examples
#' \dontrun{
#' do_coord_conversion(
#'  292500,
#'  72500
#' )
#' }
#'
#' @export


do_coord_conversion <- function(northings, eastings) {

  #define the UK coordinate projection
  uk_projection <- sf::st_crs(27700)

  #create spatial points object with coords
  north_east <- data.frame(eastings = eastings, northings = northings)
  points <- sf::st_as_sf(north_east, coords = c("eastings", "northings"), crs = uk_projection)

  #convert to latitude/longitude
  lat_long <- sf::st_transform(points, "+proj=longlat +datum=WGS84")

  #extract lat/long
  lat <- sf::st_coordinates(lat_long)[, "Y"]
  long <- sf::st_coordinates(lat_long)[, "X"]

  return(list("latitude" = lat, "longitude" = long))

}


R/api_calls.R
#' start_ukcp_observation_csv_job
#'
#' Send a query to the UKCP API for observation data and get a job (that will produce .csv outputs) started.
#' @param label Job label for your request - this will appear in the UKCP API.
#' @param API_KEY Your UKCP API key. Defaults to the `UKCP_API_KEY` environment variable if created.
#' @param identifier Identifier for the dataset you would like to query.
#' @param area_type How area you would like to cover is set up, eg a 'point', 'admin_region' or a 'bbox' (these are not the full set of options) to cover multiple grid cells.
#' @param area_details The name or coordinates (northings/eastings) of the location you are interested in. If coordinates, the the point ID or edges of your boundary. In the latter case, enter as a vector with W, S, E, N boundaries in that order.
#' @param collection The collection the queried dataset is part of.
#' @param temp_avg The time period the data is broken down by, eg 'day' or 'may' (for data for that month, 'mall' for all months).
#' @param year_range Years covered by your call. Enter as a vector in the following format: `c(min_yr, max_yr)`.
#' @param variable The measurement you are trying to extract, eg 'rainfall' or 'sun'.
#'
#' @return A list of information about your job, crucially the URL to use to get the results (when ready).
#'
#' @examples
#' \dontrun{
#'  start_ukcp_observation_csv_job(
#'   label = "ukcpR-query",
#'   identifier = "LS6_Subset_02",
#'   area_type = "point",
#'   area_details = c(292500, 72500),
#'   collection = "land-obs_5",
#'   temp_avg = "day",
#'   year_range = c(1960, 2021),
#'   variable = "rainfall"
#'  )
#'  }
#'
#' @export

start_ukcp_observation_csv_job <- function(
    label,
    API_KEY = Sys.getenv("UKCP_API_KEY"),
    identifier,
    area_type,
    area_details, #for bbox, W, S E, N boundaries, for point, eastings, northings
    collection,
    temp_avg, #could be by month (specify month or all (mall)), day etc
    year_range,
    variable
) {

  #base url
  base <- "https://ukclimateprojections-ui.metoffice.gov.uk/wps"

  #parse vector parameters
  if(is.numeric(area_details)) {
    area_details <- paste(paste0(area_details, ".0"), collapse = "%7C")
  } else {
    area_details <- stringr::str_replace(area_details, " ", "%20")
  }
  year_range <- paste(year_range, collapse = "%7C")

  #parameters
  params <- list(
    service = "WPS",
    request = "Execute",
    version = "1.0.0",
    Identifier = identifier,
    Format = "text/xml",
    Inform = "false",
    Store = "false",
    Status = "false",
    DataInputs = paste0(
      'Area=', paste0(area_type, "%7C", area_details), ';',
      'Collection=', collection, ';',
      'DataFormat=csv;',
      'JobLabel=', label, ';',
      'TemporalAverage=', temp_avg, ';',
      'TimeSlice=', year_range, ';',
      'Variable=', variable, ';'
    ),
    ApiKey = API_KEY
  )

  #make the query
  result <- httr::GET(
    url = base,
    query = params,
    timeout = httr::timeout(15)
  )

  #display error message if required
  if(result$status_code >= 400) {
    err_msg = httr::http_status(result)
    stop(err_msg)
  }

  #parse results
  result_xml <- httr::content(result, type = "text/xml")

  #extract job info
  status_url <- xml2::xml_find_first(result_xml, "//*[local-name()='ExecuteResponse']") |> xml2::xml_attr("statusLocation")
  job_id <- xml2::xml_find_first(result_xml, "//*[local-name()='JobID']") |> xml2::xml_text()
  request_url <- xml2::xml_find_first(result_xml, "//*[local-name()='JobRequestURL']") |> xml2::xml_text()

  return(list("job_id" = job_id, "req_url" = request_url, "status_url" = status_url))

}

#' start_ukcp_projection_csv_job
#'
#' Send a query to the UKCP API for projection data and get a job (that will produce .csv outputs) started.
#' @param label Job label for your request - this will appear in the UKCP API.
#' @param API_KEY Your UKCP API key. Defaults to the `UKCP_API_KEY` environment variable if created.
#' @param identifier Identifier for the dataset you would like to query.
#' @param area_details The coordinates (northings/eastings) of the location you are interested in. Enter as a vector with W, S, E, N boundaries in that order.
#' @param collection The collection the queried dataset is part of.
#' @param temp_avg The time period the data is broken down by, eg 'day' or 'may' (for data for that month, 'mall' for all months).
#' @param year_range Years covered by your call. Enter as a vector in the following format: `c(min_yr, max_yr)`. Note there may be restrictions on which years can be sent depending on which data you are requesting.
#' @param variable The measurement you are trying to extract, eg 'rainfall' or 'sun'.
#' @param climate_change_type Whether you want 'absolute' or 'anomaly' values.
#' @param ensemble_member_set Set of models used to generate projections.
#' @param scenario RCP scenario (rcp26 or rcp85) to gather projections for. Note only variable for global projections.
#' @return A list of information about your job, crucially the URL to use to get the results (when ready).
#'
#' @examples
#' \dontrun{
#'  start_ukcp_projection_csv_job(
#'   label = "ukcpR-query",
#'   identifier = "LS2_Subset_01",
#'   area_details = c(292500, 72500, 293500, 73500),
#'   collection = "land-gcm",
#'   temp_avg = "mall",
#'   year_range = c(2061, 2080),
#'   variable = "tas"
#'  )
#'  }
#'
#' @export


start_ukcp_projection_csv_job <- function(
    label,
    API_KEY = Sys.getenv("UKCP_API_KEY"),
    identifier,
    area_details, #for bbox, W, S E, N boundaries
    collection,
    temp_avg, #could be by month (specify month or all (mall)), annual etc
    year_range,
    variable,
    climate_change_type,
    ensemble_member_set,
    scenario
) {

  #error checking

  #base url
  base <- "https://ukclimateprojections-ui.metoffice.gov.uk/wps"

  #parse vector parameters
  area_details <- paste(paste0(area_details, ".0"), collapse = "%7C")
  year_range <- paste(year_range, collapse = "%7C")

  #data inputs, based on user selection
  if(identifier == "LS2_Subset_01") {

    data_inputs <- paste0(
      'Area=', paste0("bbox%7C", area_details), ';',
      'ClimateChangeType=', climate_change_type, ';',
      'Collection=', collection, ';',
      'EnsembleMemberSet=', ensemble_member_set,
      'DataFormat=csv;',
      'JobLabel=', label, ';',
      'Scenario=', scenario, ';',
      'TemporalAverage=', temp_avg, ';',
      'TimeSlice=', year_range, ';',
      'Variable=', variable, ';'
    )

  } else {

    data_inputs <- paste0(
      'Area=', paste0("bbox%7C", area_details), ';',
      'ClimateChangeType=', climate_change_type, ';',
      'Collection=', collection, ';',
      'EnsembleMemberSet=', ensemble_member_set,
      'DataFormat=csv;',
      'JobLabel=', label, ';',
      'TemporalAverage=', temp_avg, ';',
      'TimeSlice=', year_range, ';',
      'Variable=', variable, ';'
    )

  }

  #parameters
  params <- list(
    service = "WPS",
    request = "Execute",
    version = "1.0.0",
    Identifier = identifier,
    Format = "text/xml",
    Inform = "false",
    Store = "false",
    Status = "false",
    DataInputs = paste0(data_inputs),
    ApiKey = API_KEY
  )

  #make the query
  result <- httr::GET(
    url = base,
    query = params,
    timeout = httr::timeout(15)
  )

  #display error message if required
  if(result$status_code >= 400) {
    err_msg = httr::http_status(result)
    stop(err_msg)
  }

  #parse results
  result_xml <- httr::content(result, type = "text/xml")

  #extract job info
  status_url <- xml2::xml_find_first(result_xml, "//*[local-name()='ExecuteResponse']") |> xml2::xml_attr("statusLocation")
  job_id <- xml2::xml_find_first(result_xml, "//*[local-name()='JobID']") |> xml2::xml_text()
  request_url <- xml2::xml_find_first(result_xml, "//*[local-name()='JobRequestURL']") |> xml2::xml_text()

  return(list("job_id" = job_id, "req_url" = request_url, "status_url" = status_url))

}


#' download_ukcp_csv_results
#'
#' Download the results of your API query, using the status_url returned in `start_ukcp_csv_job`. Polls server once every five seconds.
#' @param status_url URL used to retrieve results from the API.
#' @param output_folder Folder (in working directory) to save .csv files to.
#' @param keep_zip Keep the original zip file (in your working directory root folder). Note the initial name will always be `temp.zip`.
#' @param API_KEY UKCP API key. Defaults to using environment variable `UKCP_API_KEY`.
#'
#' @return Downloads the .zip file held at the `status_url` and extracts the csv(s) within (users can optionally keep the .zip file).
#'
#' @examples
#' \dontrun{
#'  download_ukcp_csv_results(
#'   status_url = "https://ukclimateprojections-ui.metoffice.gov.uk/status/your-id-here",
#'   output_folder = "outputs"
#' )
#' }
#'
#' @export

download_ukcp_csv_results <- function(
    status_url,
    output_folder = "outputs",
    keep_zip = F,
    API_KEY = Sys.getenv("UKCP_API_KEY")
    ) {

  #make sure keep_zip is boolean
  if(!is.logical(keep_zip)) {
    stop("`keep_zip` must be T/TRUE or F/FALSE (unquoted)")
  }

  print("Polling server...")

  #make initial query
  result <- httr::GET(
    url = status_url,
    timeout = httr::timeout(15)
  )

  #display error message if required
  if(result$status_code >= 400) {
    err_msg = httr::http_status(result)
    stop(err_msg)
  }

  #parse results
  result_xml <- httr::content(result, type = "text/xml")

  #status as per result xml and extract child node names
  status <- xml2::xml_find_first(result_xml, "//*[local-name()='Status']")
  status_child <- xml2::xml_name(xml2::xml_children(status))

  #run the above process through a while loop until completed
  while(status_child %in% c("ProcessAccepted", "ProcessStarted")) {

    #UI
    print(paste0("Polling server...last status: ", status_child))

    #make the query
    result <- httr::GET(
      url = status_url,
      timeout = httr::timeout(15)
    )

    #display error message if required
    if(result$status_code >= 400) {
      err_msg = httr::http_status(result)
      stop(err_msg)
    }

    #parse results
    result_xml <- httr::content(result, type = "text/xml")

    #status as per result xml and extract child node names
    status <- xml2::xml_find_first(result_xml, "//*[local-name()='Status']")
    status_child <- xml2::xml_name(xml2::xml_children(status))

    #sleep for 2 seconds to avoid overwhelming API
    Sys.sleep(5)

  }

  #if result completed successfully, grab file and download...
  if(status_child == "ProcessSucceeded") {

    print("Success! Now extracting csv(s)...")

    #get the download URL from the XML response
    download_url <- result_xml |> xml2::xml_find_first("//*[local-name()='FileURL']") |> xml2::xml_text()

    #now you need to append the api key...
    download_url_w_key <- paste0(download_url, "?ApiKey=", API_KEY)

    #create the output directory if needed
    if(!dir.exists(output_folder)) {
      dir.create(output_folder)
    }

    #in the highly unlikely case that a file call temp.zip exists in their outputs folder...
    if(file.exists(paste0(output_folder, "/temp.zip"))) {

      stop("A file called `temp.zip` exists in your chosen output folder.
           This package creates a file with this name.
           Please rename, move or delete the file and try again.")

    }

    #get the csv and save it
    destination <- paste0(output_folder, "/temp.zip")
    download.file(download_url_w_key, destination, mode = "wb")

    #identify the csv file(s)
    files <- unzip(destination, list = T)
    csvs <- files$Name[grepl("csv", files$Name)]

    #for each csv identified, save in the outputs folder
    for (f in csvs) {

      unzip(destination, exdir = output_folder, files = f)

    }

    #delete the zip, if requested
    if(!keep_zip) {
      file.remove(paste0(output_folder, "/temp.zip"))
    }

    print(paste0("Finished - extracted files saved to ", output_folder))


  } else {

    #if unsuccessful, stop and inform user
    stop("Process failed; review your request and try again.")

  }

}

